exp/controllers/azuremanagedcluster_controller.go

```diff
@@ -0,0 +1,109 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	azexpv1 "sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
+
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// AzureManagedClusterReconciler reconciles a AzureManagedCluster object
+type AzureManagedClusterReconciler struct {
+	client.Client
+	Log      logr.Logger
+	Recorder record.EventRecorder
+}
+
+func (r *AzureManagedClusterReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		WithOptions(options).
+		For(&azexpv1.AzureManagedCluster{}).
+		Complete(r)
+}
+
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=azuremanagedclusters,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=azuremanagedclusters/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;clusters/status,verbs=get;list;watch
+
+func (r *AzureManagedClusterReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx := context.TODO()
```

Comment:
This hurts me a little when I see this. Should we have a time limit on a reconcile loop?

This is not so much about this PR as I'm guilty of it too, but just to bring it up.

Reply:
I usually use context.Background() here personally, I actually don't know what TODO() does?

Reply:
oh, you just mean that there's no timeout 🙂

Reply:
Precisely. It's just that we do this in all of our controllers. We should have a bounded time limit on reconcile loops.

I'll create an issue for this. It's not something that needs to change in this PR.

TODO() is just an empty context with no deadline / way to cancel

Reply:
+1 on opening an issue for this.

I saw too late this also sneaked in as a code todo in the AMP PR https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/exp/cloud/scope/machinepool.go#L151

Reply:
Makes sense, agreed we should align on this.

also ref: https://github.com/kubernetes-sigs/controller-runtime/issues/798
and less directly: https://github.com/kubernetes-sigs/controller-runtime/issues/801

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/600